server/src/main/java/org/elasticsearch/cluster/routing/BatchedRerouteService.java

```diff
@@ -95,7 +123,7 @@ public void onNoLongerMaster(String source) {
                                 pendingRerouteListeners = null;
                             }
                         }
-                        currentListeners.onFailure(new NotMasterException("delayed reroute [" + reason + "] cancelled"));
+                        currentListeners.forEach(l -> l.onFailure(new NotMasterException("delayed reroute [" + reason + "] cancelled")));
```

Comment:
Maybe use `org.elasticsearch.action.ActionListener#onFailure(java.lang.Iterable<org.elasticsearch.action.ActionListener<Response>>, java.lang.Exception)` to make this a little safer?